### ISSUE TITLE
Macro-expanded unconfigured items are gated feature checked

### ISSUE THREAD

**Issue by jseyfried** [rust-lang/rust#32840]
Macro-expanded unconfigured items are gated feature checked, but ordinary unconfigured items are not. For example, the following should compile, ...
```rust
#![cfg_attr(attr, feature(rustc_attrs))]

macro_rules! mac {
    {} => {
        #[cfg(attr)]
        mod m {
            #[rustc_error] // ... but if not(attr), this complains about the lack of the rustc_attrs feature
            fn f() {}
        }
    }
}

mac! {}
```

If we manually expand `mac! {}`, it always compiles:
```rust
#![cfg_attr(attr, feature(rustc_attrs))]

#[cfg(attr)]
mod m {
    #[rustc_error]
    fn f() {}
}
```

**Comment by jseyfried** [comment]
It looks like this was [intentional](https://github.com/rust-lang/rust/blob/3fb40c1d95dd3a14ad14d71e3ec847a0a102bd49/src/librustc_driver/driver.rs#L629-L631) (introduced in #22899), but I don't think it is a good idea.
cc @nrc

**Comment by LeoTestard** [comment]
There are arguments on #32648 about why this is not a good idea: `#[cfg]` is typically used by crates that need to build against both stable and nightly and use nightly-specific features when available (the most important example that I have in mind is syntax extensions using plugins on nightly and syntex on stable). Checking for use of gated features inside this `#[cfg]`-guarded code will break those crates, and it's not a good thing, this use case is actually useful. This check pass should be removed.
